File: Java/fundamentals/ControlFlow.java
public class ControlFlow {
    public static void main(String[] args) {
        ifElseStatements();
        switchStatement();
        switchExpression();
        ternaryOperator();
    }

    //============================================================
    //IF-ELSE STATEMENTS
    //============================================================
    public static void ifElseStatements() {
        int x = 10;
        
        if (x < 5) {
            int result = 1;
        } else if (x < 10) {
            int result = 2;
        } else if (x < 15) {
            int result = 3;
        } else {
            int result = 4;
        }
        
    }

    //============================================================
    //SWITCH STATEMENT
    //============================================================
    public static void switchStatement() {
        int day = 3;
        
        switch (day) {
            case 1:
                String d = "Monday";
                break;
            case 2:
                String d2 = "Tuesday";
                break;
            case 3:
                String d3 = "Wednesday";
                break;
            default:
                String d4 = "Unknown";
                break;
        }
        
        //Fall-through behavior (no break)
        int month = 2;
        int daysInMonth = 0;
        switch (month) {
            case 1: case 3: case 5: case 7: case 8: case 10: case 12:
                daysInMonth = 31;
                break;
            case 4: case 6: case 9: case 11:
                daysInMonth = 30;
                break;
            case 2:
                daysInMonth = 28;
                break;
            default:
                daysInMonth = 0;
        }
        
        //Switch with String (Java 7+)
        String color = "red";
        switch (color) {
            case "red":
                int code = 1;
                break;
            case "green":
                int code2 = 2;
                break;
            case "blue":
                int code3 = 3;
                break;
            default:
                int code4 = 0;
        }
    }

    //============================================================
    //SWITCH EXPRESSION (Java 12+)
    //============================================================
    public static void switchExpression() {
        int day = 3;
        
        //Switch expression with arrow syntax
        String dayName = switch (day) {
            case 1 -> "Monday";
            case 2 -> "Tuesday";
            case 3 -> "Wednesday";
            case 4 -> "Thursday";
            case 5 -> "Friday";
            case 6 -> "Saturday";
            case 7 -> "Sunday";
            case 9, 10, 11, 12 -> "Invalid";
            default -> "Invalid";
        };
        
        //Switch expression with block and yield
        String grade = "B";
        int points = switch (grade) {
            case "A" -> {
                int base = 4;
                yield base;
            }
            case "B" -> {
                int base = 3;
                yield base;
            }
            case "C" -> {
                int base = 2;
                yield base;
            }
            default -> 0;
        };
        
        //Traditional colon syntax with yield (Java 12+)
        int value = 2;
        String result = switch (value) {
            case 1:
                yield "one";
            case 2:
                yield "two";
            case 3:
                yield "three";
            default:
                yield "other";
        };
    }

    //============================================================
    //TERNARY OPERATOR
    //============================================================
    public static void ternaryOperator() {
        int a = 10;
        int b = 20;
        
        //Basic ternary: condition ? valueIfTrue : valueIfFalse
        int max = (a > b) ? a : b;
        
        //Multiple ternaries
        boolean flag1 = true;
        boolean flag2 = false;
        int result = flag1 ? (flag2 ? 1 : 2) : 3;
    }
    
}

File: Swift/fundamentals/controlFlow.swift
func controlFlowOverview() {
    ifElse()
    guardStatement()
    switchStatement()
    breakContinueLabels()
}


func ifElse() {
    let score = 85

    //Standard if/else - no parentheses required around condition (but allowed)
    if score >= 90 {
        print("A")
    } else if score >= 80 {
        print("B")
    } else {
        print("C")
    }

    //if as an expression (Swift 5.9+) - can assign result directly
    let grade = if score >= 90 { "A" } else if score >= 80 { "B" } else { "C" }
}


func guardStatement() {
    //guard is an early exit pattern - condition must be true to continue
    //else block must exit the current scope (return, throw, break, continue)
    //The guard vairable is available for the rest of the function
    func processAge(_ age: Int?) {
        guard let age = age else {
            print("No age provided")
            return //Must exit
        }
        //age is unwrapped and available for the rest of the function
        print("Age is \(age)")
    }

    //guard vs if: use guard when invalid state should exit early
    //Keeps the happy path unindented and readable
    func processScore(_ score: Int) {
        guard score >= 0 && score <= 100 else {
            print("Invalid score")
            return
        }
        //All further code assumes score is valid
        print("Score: \(score)")
    }

    processAge(23)
    processAge(nil)
    processScore(85)
}


func switchStatement() {
    let value = 3

    //Switch in Swift is exhaustive - must cover all cases or have a default
    //No fallthrough by default unlike Java/C - each case breaks automatically
    switch value {
    case 1:
        print("One")
    case 2, 3: //Multiple values in one case
        print("Two or Three")
    case 4...10: //Range matching - covered in Loops.swift
        print("Four to Ten")
    default:
        print("Something else")
    }

    //Switch on strings
    let day = "Monday"
    switch day {
    case "Monday", "Tuesday", "Wednesday", "Thursday", "Friday":
        print("Weekday")
    case "Saturday", "Sunday":
        print("Weekend")
    default:
        print("Unknown")
    }

    //Value binding in cases
    let point = (2, 0)
    switch point {
    case (0, 0):
        print("Origin")
    case (let x, 0): //Binds x to the first value, matches any point on x-axis
        print("On x-axis at \(x)")
    case (0, let y):
        print("On y-axis at \(y)")
    case (let x, let y):
        print("At \(x), \(y)")
    }

    //Where clause - adds condition to a case
    let num = 7
    switch num {
    case let n where n % 2 == 0:
        print("\(n) is even")
    case let n where n % 2 != 0:
        print("\(n) is odd")
    default:
        break
    }

    //Explicit fallthrough - opt in when needed
    switch value {
    case 3:
        print("Three")
        fallthrough //Executes next case regardless of its condition
    case 4:
        print("Three or Four")
    default:
        break
    }
}


func breakContinueLabels() {
    //break - exits current loop or switch
    for i in 1...10 {
        if i == 5 { break }
        print(i) //Prints 1 through 4
    }

    //continue - skips to next iteration
    for i in 1...5 {
        if i == 3 { continue }
        print(i) //Prints 1, 2, 4, 5
    }

    //Labeled statements - break/continue to a specific outer loop
    outer: for i in 1...3 {
        for j in 1...3 {
            if j == 2 { continue outer } //Skip to next iteration of outer loop
            print("\(i), \(j)")
        }
    }

    //return - exits function, covered in Functions.swift
    //throw - exits with error, covered in Concurrency/ErrorHandling.swift
}